WIP: Initial version of the xdc plugin support

tmichalak opened the pull request:
The initial version of the plugin is capable of reading an XDC file with the INTERNAL_VREF settings and generating a FASM file with the features that we currently have in our database that are related to the VREF settings per bank.
This plugin operates on the existing design and modifies its structure based on the content of the XDC (Xilinx Design Constraints) file.
Since the XDC file consists of Tcl commands it is read using Yosys's tcl command and processed by the new XDC commands imported to the Tcl interpreter.

The plugins for yosys have a separate repository (https://github.com/YosysHQ/yosys-plugins).
However we could simply add the plugin to the symbiflow-arch-defs.
We also need to add the concatenation step to symbiflow-arch-defs.

tmichalak committed:
examples: add example with XDC parsing for Arty

Signed-off-by: Tomasz Michalak <tmichalak@antmicro.com>

tmichalak committed:
examples: xdc: add XDC plugin

Signed-off-by: Tomasz Michalak <tmichalak@antmicro.com>

tmichalak committed:
examples: arty: add xdc plugin installation

Signed-off-by: Tomasz Michalak <tmichalak@antmicro.com>

mithro reviewed (commented):
Couple of comments;
 (a) Can we move this into its own repository? IE Can this be built as a plugin external to Yosys?
 (b) Can we split the plugin into two parts, the XDC import and the FASM output?
 (c) Can we make it clearer which parts are the vivado scripts and which parts are the Yosys scripts?

Otherwise this looks a lot nicer than I expected!

tmichalak commented:
@mithro 
> (a) Can we move this into its own repository? IE Can this be built as a plugin external to Yosys?

Yes, it can be built external to Yosys, I just added the plugin to Yosys for this PR to easier show the complete solution

> (b) Can we split the plugin into two parts, the XDC import and the FASM output?

Yes, we can have 2 separate plugins, one for XDC and another for FASM

> (c) Can we make it clearer which parts are the vivado scripts and which parts are the Yosys scripts?

I guess the confusion is with the `.tcl` scripts as currently Yosys scripts already have the `.ys` extension.
I could use `.ys.tcl` for Yosys and `.vivado.tcl` or simply `.tcl` for Vivado.

litghost commented on examples/arty/xdc-plugin/xdc.cc:
```diff
@@ -0,0 +1,212 @@
+// Coordinates of HCLK_IOI tiles associated with a specified bank
+// This is very part specific and is for Arty's xc7a35tcsg324 part
+std::unordered_map<int, std::string> bank_tiles = {
+	{14, "X1Y26"},
```
Should get this data from prjxray-db

tmichalak replied:
Done

litghost commented on examples/arty/xdc-plugin/xdc.cc:
```diff
@@ -0,0 +1,212 @@
+                        if (cell->type != ID(BANK)) continue;
+			int bank_number(cell->getParam(ID(NUMBER)).as_int());
+			int bank_vref(cell->getParam(ID(INTERNAL_VREF)).as_int());
+			*f << "HCLK_IOI3_" << bank_tiles[bank_number] <<".VREF.V_" << bank_vref << "_MV\n";
```
Need error checking on the bank number.  E.g. if bank number not found in bank_tiles, generate error message.

tmichalak replied:
The bank number is also checked in the [get_iobanks](https://github.com/SymbiFlow/yosys/pull/44/files/eddcec449fabcba42f73c13e130be6b81ff4ae0d#diff-23187a2ac32d94dbbe2c643f954f1ebfR96) function

litghost commented on examples/arty/xdc-plugin/xdc.cc:
```diff
@@ -0,0 +1,212 @@
+                        if (cell->type != ID(BANK)) continue;
+			int bank_number(cell->getParam(ID(NUMBER)).as_int());
+			int bank_vref(cell->getParam(ID(INTERNAL_VREF)).as_int());
+			*f << "HCLK_IOI3_" << bank_tiles[bank_number] <<".VREF.V_" << bank_vref << "_MV\n";
```
bank_vref has a limited set of values, need to correctly generate error if not a valid VREF?

tmichalak replied:
The valid VREF check is in the [set_property](https://github.com/SymbiFlow/yosys/pull/44/files/eddcec449fabcba42f73c13e130be6b81ff4ae0d#diff-23187a2ac32d94dbbe2c643f954f1ebfR153) function:

tmichalak commented:
Superseded by https://github.com/SymbiFlow/yosys-symbiflow-plugins/pull/1
The comments have been taken into account there.
